trace/tracestate/tracestate.go

```diff
@@ -0,0 +1,182 @@
+	keyWithVendorFormat    = `[a-z][_0-9a-z\-\*\/]{0,240}@[a-z][_0-9a-z\-\*\/]{0,13}`
+	keyFormat              = `(` + keyWithoutVendorFormat + `)|(` + keyWithVendorFormat + `)`
+	valueFormat            = `[\x20-\x2b\x2d-\x3c\x3e-\x7e]{0,255}[\x21-\x2b\x2d-\x3c\x3e-\x7e]`
+)
+
+var keyValidationRegExp = regexp.MustCompile(`^(` + keyFormat + `)$`)
+var valueValidationRegExp = regexp.MustCompile(`^(` + valueFormat + `)$`)
+
+// Tracestate represent tracing-system specific context in a list of key-value pairs. Tracestate allows different
+// vendors propagate additional information and inter-operate with their legacy Id formats.
+type Tracestate struct {
+	tracestateEntries []*TracestateEntry
+}
+
+// TracestateEntry represent one key-value pair in a list of key-value pair of Tracestate.
+// Key is an opaque string up to 256 characters printable. It MUST begin with a lowercase letter,
+// and can only contain lowercase letters a-z, digits 0-9, underscores _, dashes -, asterisks *, and
+// forward slashes /.
+//
+// Value is an opaque string up to 256 characters printable ASCII RFC0020 characters (i.e., the
+// range 0x20 to 0x7E) except comma , and =.
+type TracestateEntry struct {
+	key   string
+	value string
+}
+
+// Key returns the key of TracestateEntry
+func (te *TracestateEntry) Key() string {
+	return te.key
+}
+
+// Value returns the value of TracestateEntry
+func (te *TracestateEntry) Value() string {
+	return te.value
+}
+
+// Get retrieves value for a given key from Tracestate ts.
+// If the key is not found then false is returned with the value "" is returned.
+// If the key is found then true is returned with its value.
+func (ts *Tracestate) Get(key string) (string, bool) {
+	if len(ts.tracestateEntries) == 0 {
+		return "", false
+	}
+	for _, entry := range ts.tracestateEntries {
+		if entry.key == key {
+			return entry.value, true
+		}
+	}
+	return "", false
+}
+
+// TraceEntries returns a slice of TracestateEntry.
+func (ts *Tracestate) TraceEntries() []*TracestateEntry {
+	return ts.tracestateEntries
+}
+
+func (ts *Tracestate) remove(key string) {
+	for index, entry := range ts.tracestateEntries {
+		if entry.key == key {
+			ts.tracestateEntries = append(ts.tracestateEntries[:index], ts.tracestateEntries[index+1:]...)
+			break
+		}
+	}
+}
+
+func (ts *Tracestate) add(entry *TracestateEntry) error {
+	ts.remove(entry.Key())
+	if len(ts.tracestateEntries) >= maxKeyValuePairs {
+		return fmt.Errorf("Set failed: reached maximum key/value pairs limit of %d", maxKeyValuePairs)
+	}
+	ts.tracestateEntries = append([]*TracestateEntry{entry}, ts.tracestateEntries...)
+	return nil
+}
+
+func isValid(key, value string) bool {
+	return keyValidationRegExp.MatchString(key) &&
+		valueValidationRegExp.MatchString(value)
+}
+
+// NewTracestateEntry creates a TracestateEntry object with given key and value.
+// It returns error if either key or value is invalid.
+func NewTracestateEntry(key, value string) (*TracestateEntry, error) {
```

Comment:
You don't need that if you make Entry

```
type Entry struct{ Key, Value string }
```

Then, user can construct entries as Entry{Key: key, Value: value}